### PR TITLE
add assigned seqno before sorting.

### PR DESCRIPTION
**Pull request by jijiew**
Summary:
bug fix for the issue/6245
https://github.com/facebook/rocksdb/issues/6245

Test Plan:
test added in test file

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: db/external_sst_file_basic_test.cc [hunk 1]
```diff
@@ -1542,6 +1542,43 @@ TEST_F(ExternalSSTFileBasicTest, OverlappingFiles) {
...
+  ExternalSstFileInfo file1_info;
+  Status s = sst_file_writer.Finish(&file1_info);
+  ASSERT_OK(s) << s.ToString();
+
+  // Current file size should be non-zero after success write.
+  ASSERT_GT(sst_file_writer.FileSize(), 0);
+
+  IngestExternalFileOptions ifo;
+  ifo.allow_global_seqno = true;
+  ifo.write_global_seqno = true;
```
Comment (mrambacher) [inline db/external_sst_file_basic_test.cc:1574]: Is there a reason you need "write_global_seqno=true".  Is it part of this fix/required?  Comments might help.

File: db/external_sst_file_basic_test.cc [hunk 2]
```diff
@@ -1542,6 +1542,43 @@ TEST_F(ExternalSSTFileBasicTest, OverlappingFiles) {
...
+  s = db_->IngestExternalFile({file1}, ifo);
```
Comment (mrambacher) [inline db/external_sst_file_basic_test.cc:1575]: Do you need to check s?  Should there be an ASSERT_OK(s) here?

File: db/external_sst_file_ingestion_job.cc [hunk 3]
```diff
@@ -622,7 +622,10 @@ Status ExternalSstFileIngestionJob::GetIngestedFileInfo(
     if (key.sequence != 0) {
       return Status::Corruption("External file has non zero sequence number");
     }
-    file_to_ingest->smallest_internal_key.SetFrom(key);
+
+    SequenceNumber assigned_seqno = versions_->LastSequence();
+
+    file_to_ingest->smallest_internal_key.Set(key.user_key, assigned_seqno, ValueType::kTypeValue);
```
Comment (mrambacher) [inline db/external_sst_file_ingestion_job.cc:628]: Shouldn't the last arg be key.type and not ValueType::kTypeValue??

File: db/external_sst_file_ingestion_job.cc [hunk 4]
```diff
@@ -633,7 +636,8 @@ Status ExternalSstFileIngestionJob::GetIngestedFileInfo(
     if (key.sequence != 0) {
       return Status::Corruption("External file has non zero sequence number");
     }
-    file_to_ingest->largest_internal_key.SetFrom(key);
+
+    file_to_ingest->largest_internal_key.Set(key.user_key, assigned_seqno, ValueType::kTypeValue);
```
Comment (mrambacher) [inline db/external_sst_file_ingestion_job.cc:640]: Same as above.

**Review by mrambacher — COMMENTED** [review]
I think a few more test cases might be in order:
- What happens if you ingest when data is not flushed?
- What happens if you Get after ingest without the compact?
- What happens if there are more keys in the ingested file (e.g. {j, k}, {k, l}, {j,k,l} where the key in question is not always the smallest and/or largest

Generally, this looks okay to me, but I am not an expert in this area....

File: db/external_sst_file_ingestion_job.cc [hunk 5]
```diff
@@ -622,7 +622,10 @@ Status ExternalSstFileIngestionJob::GetIngestedFileInfo(
     if (key.sequence != 0) {
       return Status::Corruption("External file has non zero sequence number");
     }
-    file_to_ingest->smallest_internal_key.SetFrom(key);
+
+    SequenceNumber assigned_seqno = versions_->LastSequence();
```
Comment (riversand963) [inline db/external_sst_file_ingestion_job.cc:626]: Maybe I am missing some context from the original bug, but I think we may later call `AssignLevelAndSeqnoForIngestedFile(...)` to compute `assigned_seqno` which does not have to be the current latest seqno in the db.

**Comment by jijiew** [comment]
> I think a few more test cases might be in order:
> 
> * What happens if you ingest when data is not flushed?
> * What happens if you Get after ingest without the compact?
> * What happens if there are more keys in the ingested file (e.g. {j, k}, {k, l}, {j,k,l} where the key in question is not always the smallest and/or largest
> 
> Generally, this looks okay to me, but I am not an expert in this area....

thanks for the review, more test added.

File: db/external_sst_file_basic_test.cc (hunk 1 above)
Comment (jijiew) [reply db/external_sst_file_basic_test.cc:1574]: removed, not necessary.

File: db/external_sst_file_basic_test.cc (hunk 2 above)
Comment (jijiew) [reply db/external_sst_file_basic_test.cc:1575]: Done.

File: db/external_sst_file_ingestion_job.cc (hunk 3 above)
Comment (jijiew) [reply db/external_sst_file_ingestion_job.cc:628]: hmm, I Could not find the key.type, the value type is defined here:
https://github.com/facebook/rocksdb/blob/711881bc257d761fd44924686069cb5a0cea580b/db/dbformat.h

File: db/external_sst_file_ingestion_job.cc (hunk 4 above)
Comment (jijiew) [reply db/external_sst_file_ingestion_job.cc:640]: same above, I could not find the key.type.

File: db/external_sst_file_ingestion_job.cc (hunk 5 above)
Comment (jijiew) [reply db/external_sst_file_ingestion_job.cc:626]: just did a bit of refactoring to include that part.

**Review by ajkr — COMMENTED** [review]
A couple things to do:

- run `make format`
- do you mind mentioning this under the unreleased bug fixes section in the HISTORY.md file?

A general comment on the approach - it prevents newly ingested files from being written with incorrect smallest/largest key metadata, which is a good step forward. Reading from a DB with pre-existing ingested files could still be problematic. That appears difficult to fix. Perhaps we can have a consistency check that `FileMetaData::smallest` and `FileMetaData::largest` have seqnums in [`FileMetaData::smallest_seqno`, `FileMetaData::largest_seqno`]. This can be done later.

**Pull request closed without merging**